### PR TITLE
🐛 enable managedAPIServerLoadBalancer w/ existing network

### LINKED ISSUE #547: Unable to use an existing network and `managedAPIServerLoadBalancer: true`
/kind bug

**What steps did you take and what happened:**

I created a cluster using:
 - an existing OpenStack network/subnet
 - `managedAPIServerLoadBalancer` enabled

The manifest I use is:

```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
kind: OpenStackCluster
metadata:
  name: openstack
  namespace: default
spec:
  apiServerLoadBalancerAdditionalPorts:
  - 22
  apiServerLoadBalancerFloatingIP: 192.168.50.180
  apiServerLoadBalancerPort: 6443
  cloudName: openstack
  cloudsSecret:
    name: openstack-cloud-config
    namespace: default
  disablePortSecurity: true
  disableServerTags: true
  dnsNameservers:
  - 192.168.50.254
  externalNetworkId: dd02a459-0e4a-444d-9610-355709acbf1a
  managedAPIServerLoadBalancer: true
  network:
    name: "private-net"
  subnet:
    name: "private-subnet"
  useOctavia: false
```

`192.168.50.180` is an existing _unassigned_ floating address. And `private-net/subnet` are existing Network/Subnet.

Then no load balancer are created and thus the control plane node cannot initialize.

**What did you expect to happen:**

That a load balancer is created and that the control plan is initialized.

**Anything else you would like to add:**

If I set nodeCIDR instead of specifying and existing network and subnet. The setup works properly and a load balancer is created.

It seems that https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/controllers/openstackcluster_controller.go#L251 is missing a similar instruction `reconcileLoadBalancer` as the `else` case.

Note that if I delete the Cluster object then the `capo-controller-manager` crashes with:

```
E0420 12:06:03.224464       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 396 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1524ee0, 0x2427830)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1524ee0, 0x2427830)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
sigs.k8s.io/cluster-api-provider-openstack/pkg/cloud/services/loadbalancer.(*Service).DeleteLoadBalancerMember(0xc0006b2980, 0xc0005d21e0, 0x11, 0xc0004e2fc0, 0xc0004e2b40, 0xc0003a2500, 0x0, 0x0)
	/workspace/pkg/cloud/services/loadbalancer/loadbalancer.go:420 +0x1dd
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).reconcileDelete(0xc0003f6ba0, 0x191a680, 0xc000080080, 0x19292e0, 0xc000724330, 0xc000526230, 0xc0004e2fc0, 0xc0004e2b40, 0xc0006a7980, 0xc0003a2500, ...)
	/workspace/controllers/openstackmachine_controller.go:216 +0x759
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).Reconcile(0xc0003f6ba0, 0xc0005061f9, 0x7, 0xc00047e900, 0x1d, 0xc0004ddc00, 0x0, 0x0, 0x0)
	/workspace/controllers/openstackmachine_controller.go:146 +0x9e5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0002220c0, 0x15780e0, 0xc0001c8620, 0xc000630500)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.1/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0002220c0, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.1/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0002220c0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.1/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0005e13b0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0005e13b0, 0x3b9aca00, 0x0, 0x1, 0xc000602000)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc0005e13b0, 0x3b9aca00, 0xc000602000)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.1/pkg/internal/controller/controller.go:193 +0x328
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x13c7a3d]
```

**Environment:**

- Cluster-api version: 

```
$ clusterctl version
clusterctl version: &version.Info{Major:"0", Minor:"3", GitVersion:"v0.3.3", GitCommit:"ecff70af1b839c4086335234d88b1b8c00c3383c", GitTreeState:"clean", BuildDate:"2020-03-27T22:25:19Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```

- Minikube/KIND version:

```
$ kind version
kind v0.7.0 go1.13.6 linux/amd64
```

- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by gdetal**
**What this PR does / why we need it**:

Using `managedAPIServerLoadBalancer: true` in the `OpenStackCluster` configuration suggest that a load balancer will be created. However, when specifying an existing Openstack network instead of using `nodeCidr`, a load balancer is *not* created.

This PR make sure that:
 - the load balancer is created;
 - the controller do not try to remove the existing network.

**Which issue(s) this PR fixes** *
fixes #547 

**Special notes for your reviewer**:

- I had to refactor `reconcileNormal` as the linter was complaing about a cyclomatic complexity above 30 after my change.

**Release note**:
```release-note

```

### CODE REVIEW

**Comment by gdetal** [comment]
CLA check

**Comment by sbueringer** [comment]
> Hey folks, I don't work in this project anymore. Would you mind removing me from the reviewers list? Thanks

@flaper87 Yup, no problem

File: controllers/openstackcluster_controller.go [hunk 1]
```diff
@@ -158,7 +160,8 @@ func (r *OpenStackClusterReconciler) reconcileDelete(ctx context.Context, log lo
 		log.Info("OpenStack router deleted successfully")
 	}
 
-	if openStackCluster.Status.Network != nil {
+	// if NodeCIDR was not set, no network was created.
+	if openStackCluster.Status.Network != nil && openStackCluster.Spec.NodeCIDR != "" {
```
Comment (sbueringer) [inline controllers/openstackcluster_controller.go:164]: Just that I get it correctly. We were using an already existing network and that's why we don't want to delete it via CAPO?

File: controllers/openstackcluster_controller.go [hunk 2]
```diff
@@ -239,40 +223,75 @@ func (r *OpenStackClusterReconciler) reconcileNormal(ctx context.Context, log lo
...
 	err = networkingService.ReconcileSecurityGroups(clusterName, openStackCluster)
 	if err != nil {
-		return reconcile.Result{}, errors.Errorf("failed to reconcile security groups: %v", err)
+		return errors.Errorf("failed to reconcile security groups: %v", err)
+	}
+
+	return nil
+}
+
+func (r *OpenStackClusterReconciler) reconcileNormal(ctx context.Context, log logr.Logger, patchHelper *patch.Helper, cluster *clusterv1.Cluster, openStackCluster *infrav1.OpenStackCluster) (ctrl.Result, error) {
```
Comment (sbueringer) [inline controllers/openstackcluster_controller.go:272]: Can you reorder the methods please and put reconcileNormal above reconcileNetworkComponents?  (imho just a lot easier to read it "top down")

File: controllers/openstackcluster_controller.go [hunk 3]
```diff
@@ -239,40 +223,75 @@ func (r *OpenStackClusterReconciler) reconcileNormal(ctx context.Context, log lo
...
-			return reconcile.Result{}, errors.Errorf("failed to find only one subnet (result: %v): %v", subnetList, err)
+			return errors.Errorf("failed to find only one subnet (result: %v): %v", subnetList, err)
 		}
 		openStackCluster.Status.Network.Subnet = &infrav1.Subnet{
 			ID:   subnetList[0].ID,
 			Name: subnetList[0].Name,
 			CIDR: subnetList[0].CIDR,
 		}
+
+		if openStackCluster.Spec.ManagedAPIServerLoadBalancer {
```
Comment (sbueringer) [inline controllers/openstackcluster_controller.go:237]: I think we can just move this part below the if/else so we don't have to duplicate it?

**Comment by sbueringer** [comment]
@gdetal just fyi the openlab test has currently an open issue unrelated to your PR, but it's optional for merge (https://github.com/theopenlab/openlab/issues/504)

File: controllers/openstackcluster_controller.go (hunk 1 above)
Comment (gdetal) [reply controllers/openstackcluster_controller.go:164]: Yes that is correct. In my setup I use an existing network that is shared with other VMs. I think it make sense to not clean resources that are not allocated by the controller.

File: controllers/openstackcluster_controller.go (hunk 2 above)
Comment (gdetal) [reply controllers/openstackcluster_controller.go:272]: Will do !

File: controllers/openstackcluster_controller.go (hunk 3 above)
Comment (gdetal) [reply controllers/openstackcluster_controller.go:237]: Yes, It was basically a copy/paste.

File: controllers/openstackcluster_controller.go (hunk 1 above)
Comment (sbueringer) [reply controllers/openstackcluster_controller.go:164]: Yup makes sense

**Comment by sbueringer** [comment]
/approve
recheck

**Comment by sbueringer** [comment]
/assign 
@jichenjc can you also take a look please?

**Comment by gdetal** [comment]
Thanks for the review, I still have that CLA issue ongoing... I signed the document but still the bot  didn't update the label... @k8s-ci-robot

**Comment by sbueringer** [comment]
Hm the recheck command should work (https://prow.k8s.io/plugins) Looks like something wrong with either your CLA or the CLA system / bot

**Comment by gdetal** [comment]
@sbueringer that is what I thought, I'll create a ticket on the linuxfoundation support website.

**Pull request merged**